src/SearchIndex.ts

```diff
@@ -0,0 +1,328 @@
+                resolve(data.toString('hex'))
+            }
+        });
+
+        hash.write(data);
+        hash.end();
+    })
+}
+
+async function getManifestsFromS3(bucketName: string, prefix: string): Promise<Manifest[]> {
+    const s3 = new S3({apiVersion: '2006-03-01'})
+    const result: S3.Types.ListObjectsV2Output = await util.promisify(s3.makeUnauthenticatedRequest.bind(s3, 'listObjectsV2', {
+        Bucket: bucketName,
+        Prefix: prefix
+    }))()
+
+    if (result.IsTruncated) {
+        // This would indicate something awry, since we shouldn't
+        // ever have more than 1000 properties. And if we ever did,
+        // everything would need to be rearchitected.
+        throw new Error('Got truncated response from S3')
+    }
+
+    const manifests = []
+    for (const bucketEntry of (result.Contents || [])) {
+        if (bucketEntry.Size === 0) {
+            log.error(new Error(`Got empty file: "${bucketEntry.Key}"`))
+            continue
+        }
+
+        assert.ok(bucketEntry.Key)
+
+        const matches = bucketEntry.Key.match(/([^/]+).json$/)
+        if (matches === null) {
+            log.error(new Error(`Got weird filename in manifest listing: "${bucketEntry.Key}"`))
+            continue
+        }
+
+        const searchProperty = matches[1]
+        const data: S3.Types.GetObjectOutput = await util.promisify(s3.makeUnauthenticatedRequest.bind(s3, 'getObject', {
+            Bucket: bucketName,
+            Key: bucketEntry.Key
+        }))()
+
+        assert.ok(data.Body)
+        assert.ok(data.LastModified)
+
+        const body = data.Body.toString('utf-8')
+        const hash = await generateHash(body)
+        const parsed = JSON.parse(body)
+        manifests.push({
+            manifest: parsed,
+            lastModified: data.LastModified,
+            manifestRevisionId: hash,
+            searchProperty: searchProperty
+        })
+    }
+
+    return manifests
+}
+
+function getManifestsFromDirectory(prefix: string): Promise<Manifest[]> {
+    return new Promise((resolve, reject) => {
+        const manifests: Manifest[] = []
+
+        dive(prefix, async (err: Error | null, path: string, stats: fs.Stats) => {
+            if (err) { reject(err) }
+            const matches = path.match(/([^/]+).json$/)
+            if (!matches) { return }
+            const searchProperty = matches[1]
+            const data = fs.readFileSync(path, {encoding: 'utf-8'})
+            const parsed = JSON.parse(data)
+            const hash = await generateHash(data)
+
+            manifests.push({
+                manifest: parsed,
+                lastModified: stats.mtime,
+                manifestRevisionId: hash,
+                searchProperty: searchProperty
+            })
+        }, () => {
+            resolve(manifests)
+        })})
+}
+
+async function getManifests(manifestSource: string): Promise<Manifest[]> {
+    const parsedSource = manifestSource.match(/((?:bucket)|(?:dir)):(.+)/)
```

Comment:
May be helpful to describe the format we're looking for in `manifestSource` somewhere.